### PR TITLE
Ledger validator - use traversal utils

### PR DESCRIPTION
**Pull request by tsvisabo**
- Refactor `tangle_traversal_dfs_to_genesis` - enable storing `analyzed_hashes` + remove unused param
- `get_latest_delta` - remove `counted_hashes` - (`bundle`  arg was accumulating transactions from every scanned bundle and i guess that's the reason for the additional counted_hashes set)
- Use traversal in `get_latest_delta`

### CODE REVIEW

File: common/model/bundle.c [hunk 1]
```diff
@@ -18,8 +18,10 @@ void bundle_transactions_new(bundle_transactions_t **const bundle) {
 }
 
 void bundle_transactions_free(bundle_transactions_t **const bundle) {
-  utarray_clear(*bundle);
-  utarray_free(*bundle);
+  if (bundle && *bundle){
```
Comment (th0br0) [inline common/model/bundle.c]: :+1:

File: consensus/transaction_solidifier/transaction_solidifier.h [hunk 2]
```diff
@@ -36,8 +38,8 @@ typedef struct transaction_solidifier_s {
 } transaction_solidifier_t;
 
 extern retcode_t iota_consensus_transaction_solidifier_init(
-    transaction_solidifier_t *const ts, tangle_t *const tangle,
-    requester_state_t *const requester);
+    transaction_solidifier_t *const ts, iota_consensus_defs_t *const defs,
```
Comment (th0br0) [inline consensus/transaction_solidifier/transaction_solidifier.h:41]: Could you add comments to the methods here detailing what they do? 
at least the check_solidity & such, init/start/stop/destroy arguably don't need that

File: consensus/utils/tangle_traversals.c [hunk 3]
```diff
@@ -9,13 +9,15 @@
...
+retcode_t tangle_traversal_dfs_to_genesis(
+    tangle_t *const tangle, tangle_traversal_functor func,
+    flex_trit_t *const entry_point, flex_trit_t *const genesis_hash,
+    hash243_set_t *const analyzed_hashe_param, void *data) {
   retcode_t ret = RC_OK;
   hash243_stack_t non_analyzed_hashes = NULL;
-  hash243_set_t analyzed_hashes = NULL;
+  hash243_set_t analyzed_hashes_local = NULL;
+  hash243_set_t *analyzed_hashes =
+      analyzed_hashe_param ? analyzed_hashe_param : &analyzed_hashes_local;
```
Comment (th0br0) [inline consensus/utils/tangle_traversals.c]: typo `hashe_param`

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: consensus/transaction_solidifier/transaction_solidifier.h (hunk 2 above)
Comment (tsvisabo) [reply consensus/transaction_solidifier/transaction_solidifier.h:41]: That will be helpful, it should be done in a different PR IMO, though i was thinking that next PR will even eliminate that function as we discussed

File: consensus/ledger_validator/ledger_validator.c [hunk 4]
```diff
@@ -126,100 +128,99 @@ static retcode_t build_snapshot(ledger_validator_t const *const lv,
...
+                                          bool *should_stop) {
   retcode_t ret = RC_OK;
-  hash243_stack_t non_analyzed_hashes = NULL;
-  hash243_set_t counted_hashes = NULL;
-  iota_transaction_t tx_bundle = NULL;
-  DECLARE_PACK_SINGLE_TX(tx, tx_ptr, pack);
-  struct _trit_array curr_hash = {NULL, NUM_TRITS_HASH, FLEX_TRIT_SIZE_243, 0};
-
   bool valid_bundle = false;
+  struct _trit_array curr_hash = {NULL, NUM_TRITS_HASH, FLEX_TRIT_SIZE_243, 0};
```
Comment (thibault-martinez) [inline consensus/ledger_validator/ledger_validator.c]: Can you add member designators here ?

**Review by th0br0 — APPROVED** [review]

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**